### DISCUSSION TITLE
How to use generic traits? [Question]

### DISCUSSION THREAD

**Discussion by BA8F0D39** [arrayfire/arrayfire-rust#235] [Q&A] [answered]
```
pub fn addnum(a: &arrayfire::Array<arrayfire::FloatingPoint>
	,b: &arrayfire::Array<arrayfire::FloatingPoint>) ->  arrayfire::Array<arrayfire::FloatingPoint> {
  a + b
}
```

I want to make a function that accepts the trait arrayfire::FloatingPoint and produces arrayfire::FloatingPoint. What modifications do I need to make it compile?

**Comment by 9prady9** [comment]
@BA8F0D39 I am not sure I understand your use case, the trait you mentioned is more like a `contract checking` so that only acceptable `Array` objects(Array is generic struct) are passed to the relevant functions. 

Is this what you are trying to achieve ?
```rust
pub fn addnum(a: &arrayfire::Array<T>,b: &arrayfire::Array<T>) ->  arrayfire::Array<T> 
where T: FloatingPoint {
  a + b
}
```

**Comment by BA8F0D39** [comment]
I am new to rust and the rust compiler says cannot find type `T` in this scope.
How would you define the trait T in this case to use arrayfire::FloatingPoint?

**Comment by 9prady9** [comment, marked as answer]
In that case, you should look at https://doc.rust-lang.org/rust-by-example/generics.html

`fn foo<T>(arg: T) { ... }` that is how you define a generic function. In this case, the generic type `T` is what reused for defining the generic type `Array<>` that is in turn used a arguments to the function.
